pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -213,20 +248,169 @@ func (c *MCDefaultRouteController) processNextWorkItem() bool {
 	}
 	defer c.queue.Done(obj)
 
-	if k, ok := obj.(string); !ok {
+	key, ok := obj.(string)
+	if !ok {
 		c.queue.Forget(obj)
 		klog.InfoS("Expected string in work queue but got", "object", obj)
 		return true
-	} else if err := c.syncMCFlows(); err == nil {
-		c.queue.Forget(k)
+	}
+
+	syncFn := func() error {
+		if c.wireGuardConfig != nil {
+			if err := c.reconcileWireGuard(); err != nil {
+				return err
+			}
+		}
+		return c.syncMCFlows()
+	}
+	if err := syncFn(); err == nil {
+		c.queue.Forget(key)
 	} else {
 		// Put the item back on the workqueue to handle any transient errors.
-		c.queue.AddRateLimited(k)
-		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", key)
 	}
 	return true
 }
 
+// reconcileWireGuard reconciles WireGuard configurations in following way:
+//  1. If the current Node is the Multi-cluster Gateway Node, controller will try to initialize corresponding WireGuard
+//     configuration and route on the host, then add all existing ClusterInfoImports as WireGuard peers.
+//  2. If the current Node is not Multi-cluster Gateway Node, controller will try to clean up WireGuard.
+//
+// Note: MCDefaultRouteController runs only one worker to process Gateway and ClusterInfoImport. So we do not need
+// any synchronization mechanism.
+func (c *MCDefaultRouteController) reconcileWireGuard() error {
+	gateway, err := c.getActiveGateway()
+	if err != nil {
+		return err
+	}
+
+	amIGateway := gateway != nil && gateway.Name == c.nodeConfig.Name
+	if !amIGateway {
```

Comment:
Given both `initializeWireGuard `and `cleanUpWireGuard` can fail, and `initializeWireGuard `does not support reentrance, I think you should still do:

```
		if c.wireGuardClient != nil && (!amIGateway || !c.wireGuardInitialized) {
			if err := c.cleanUpWireGuard(); err != nil {
				return err
			}
			c.wireGuardInitialized = false
		}
```

Reply:
Then you just need to return nil in this block (`if !amIGateway`)

Reply:
Done.